server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestResolveClusterAction.java

```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.rest.action.admin.indices;
+
+import org.elasticsearch.action.admin.indices.resolve.ResolveClusterAction;
+import org.elasticsearch.client.internal.node.NodeClient;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.rest.BaseRestHandler;
+import org.elasticsearch.rest.RestRequest;
+import org.elasticsearch.rest.Scope;
+import org.elasticsearch.rest.ServerlessScope;
+import org.elasticsearch.rest.action.RestToXContentListener;
+
+import java.io.IOException;
+import java.util.List;
+
+import static org.elasticsearch.rest.RestRequest.Method.GET;
+
+@ServerlessScope(Scope.PUBLIC)
+public class RestResolveClusterAction extends BaseRestHandler {
+
+    @Override
+    public String getName() {
+        return "resolve_cluster_action";
+    }
+
+    @Override
+    public List<Route> routes() {
+        return List.of(new Route(GET, "/_resolve/cluster/{name}"));
+    }
+
+    @Override
+    protected BaseRestHandler.RestChannelConsumer prepareRequest(RestRequest request, NodeClient client) throws IOException {
+        String[] indexExpressions = Strings.splitStringByCommaToArray(request.param("name"));
+        ResolveClusterAction.Request resolveRequest = new ResolveClusterAction.Request(indexExpressions);
+        return channel -> client.admin().indices().resolveCluster(resolveRequest, new RestToXContentListener<>(channel));
```

Comment:
Can we make this cancellable (i.e. use a `RestCancellableNodeClient` here, and check the task for cancellation at various points during execution) please?

Reply:
Fixed in next push.

Reply:
@dct - I did a first attempt at adding cancel-ability to this endpoint in this commit: https://github.com/elastic/elasticsearch/pull/102726/commits/bcd243704f78e2039a008e8f4713f9bf598b7adf

However, there are two problems I'm seeing that I'd like to pointers on how to solve.

First, the cancellation request is not being propagated to the remote clusters. (I tested this by adding in some loops with sleeps in the code that executes remotely and after cancelling the task on the primary cluster, only the primary cluster sees the cancellation, not the remote actions.) I think I might need to set parent task somewhere, but it's unclear to me where. I don't see any other TransportActions that do it explicitly. It seems like it's supposed to be handled in the lower Transport layer?

I also tried following the field-caps model of registering a separate [NodeTransportHandler](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/action/fieldcaps/TransportFieldCapabilitiesAction.java#L500) in this class (not in the pushed commit), but that didn't work - that code does not execute on the remote cluster.

Second, when I cancel the task (via `POST _tasks/<id>/_cancel`), the task does get cancelled but it never disappears from the `GET _tasks` list, so some cleanup/shutdown code must be necessary. Also, I'm not completely sure what to do when an action detects that a task is cancelled. Should I just call `task.ensureNotCancelled()`, and just the `TaskCancelledException` be returned?  I'm not clear what should be done on the primary coordinator when it detects it nor what to do on the remote node when it is detected.

Is there are example TransportAction that is cross-cluster and handles these aspects correctly?

Reply:
> I think I might need to set parent task somewhere, but it's unclear to me where.

Yes that's right, you need to call `remoteRequest.setParentTask()`. That's important for things like APM too, and should be enough to propagate the cancellation to remote clusters.

> some cleanup/shutdown code must be necessary

That suggests you're not completing the listener. You have to do that, the task won't be unregistered until you do. Typically this would be with a `TaskCancelledException` (either using `task.ensureNotCancelled()` or else `task.notifyIfCancelled()`).

Reply:
Thanks @dct. 

I have one of the two fixed.

In [this commit](https://github.com/elastic/elasticsearch/pull/102726/commits/d1f6d81226d3416685abaae358a936369963675d), I got the unregistration working so it disappears from the task list when it completes.

However, in this [follow-on commit](https://github.com/elastic/elasticsearch/pull/102726/commits/1244453470df45c6ea4554e2de20998e39024e84), where I added `remoteRequest.setParentTask()`, the cancellation is still not being received on the remote clusters (I tested by adding in loops with sleeping and checking `resolveClusterTask.isCancelled()` in this part of the code, but it never gets a cancellation=true status). Hopefully, I missed something simple, but I don't know how to make it work at present.